connectors/source.py

```diff
@@ -341,14 +341,15 @@ class BaseDataSource:
     service_type = None
     support_incremental_sync = False
 
-    def __init__(self, configuration):
+    def __init__(self, configuration, extraction_config=None):
```

Comment:
Can we declare this as a breaking change in release? Realistically we don't have a lot of customers, most likely just a couple of people tried to make a connector at this point and they probably would have to have a good coding knowledge.

I'm not sure about adding more stuff to base class - composition is more convenient and expressive IMO - it's easier to see what you've got checking parameters of the method VS going to the base class and reading methods in it.

Also I think it's good to discuss and try to introduce breaking changes and handle it really well now since we don't have a lot of people - to do it before we have too many people around and see how bad the change could be.

How do you feel about it @seanstory ?

Reply:
Several thoughts:

- we're not yet mature enough to worry about not changing the signature; we've done it for DSL and incremental sync, and evolving the constructor is something we should be comfortable doing
- that class signature is our contract for folks creating new connectors, I initially isolated it there for that reason, and we can communicate in the CHANGE doc how it evolves
- later on the road, breaking changes can be dealt with by using deprecation warnings and a phase-out release that support both versions. 

> Composition is more convenient and expressive IMO

Maybe?  It boils down to what this class needs to get initialized and function properly instead of building helpers. Passing along the configuration seems OK to me. Maybe we could pass the whole configuration, so we're future-proof for the source class to get new options in the future

Reply:
Actually returning to original point of change - IMO this change is framework-specific and should not actually affect connectors.

The classes that handle ingestion (Fetcher? Bulker? SyncJobRunner?) should take care of it, while connectors should be agnostic to it.

So can we instead push the logic outside of connectors and make connectors return expected data - e.g. file streams?

Reply:
> connectors should be agnostic to it

But the SPO source class needs to grab that config. how would you implement it so the source class can read the configuration?

Reply:
> Can we declare this as a breaking change in release?

No, I do not think we can. 8.8.0 GA'd our product, and we now need to hold ourselves to a higher standard. Even if we don't have many customers, if we give them a really bad first impression, it will be harder for us to build confidence in our community.

> composition is more convenient and expressive IMO - it's easier to see what you've got checking parameters of the method VS going to the base class and reading methods in it.

We could switch to a Builder pattern that allows more flexible construction? Or introduce a second constructor that takes kwargs so that it can grow organically?

> we're not yet mature enough to worry about not changing the signature; we've done it for DSL and incremental sync

AFAIU, we haven't changed signatures, we've added new functions, with default no-ops in the base class. So those changes are backwards compatible. If I've misunderstood, we should revisit those and ensure we're not breaking backwards compatibility.

Reply:
>AFAIU, we haven't changed signatures, we've added new functions, with default no-ops in the base class. So those changes are backwards compatible.

That is correct.

The breaking change is only for developers who trying to build custom connectors. Is the custom connector supported by us? I didn't find it in the official guide, but it's labeled `Beta` in UI:

![image](https://github.com/elastic/connectors-python/assets/54903978/5da5ee51-5798-4c43-8dea-f825ee6e75c8)

Reply:
Here are a couple of my thoughts: 

From what I see in the configuration the extraction service is a shared service so we're not going to have multiple ones (or per connector) in this case I don't see a need to pass the service configuration over to all connectors. 

A connector sees the extraction service as a 3rd party service (a black box) so I do not expect them to manage HTTP connections or error handling. Instead, connectors send files and receive the extracted content. Something like: 

```
import extraction_service

for doc in docs: 
   if is_file(doc):
    text = ExtractionService.extract(doc)
....
```

At the same time, `ExtractionService` is an abstraction that gets initiated when the connector's service starts and takes care of the connection pool, errors, etc. Our customers can just start using it if they want by importing a required file and don't care about the configuration.